File: Old Contest and Practice Problems/coinstokens.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    int nums[1001];
    nums[0]=0;
    for(int i=1;i<1001;i++)
    {
        int sum=0;
        for(int j=1;j<(int)(sqrt(i)+1);j++)
        {
            if (i%j==0)
            {
                if (j!=i/j) sum+=(j+(i/j));
                else sum+=j;
            }
        }
        sum-=(i+1);
        nums[i]=max(sum,i);
    } //Code to generate the list of return values for each coin.
  
  int val[1001];
  val[0] = 0;
  for (int i = 1; i<=1000; i++) 
    { 
        int max_val = -1; 
        for (int j = 0; j < i; j++) {max_val = max(max_val, nums[j+1] + val[i-j-1]);} 
        val[i] = max_val; 
    }
  
  int q;cin>>q;
  
  int n;
  for(int i=0;i<q;i++)
  {
      cin>>n;
      cout<<val[n]<<endl;
  }
  return 0;
}

File: Old Contest and Practice Problems/coinstokens.py
nums=[0]
for i in range(1,1001):
    sum=0
    for j in range(1,int(i**(1/2)+1)):
        if (i%j)==0: 
            if j!=(i/j): sum+=(j+(i/j))
            else: sum+=j
    sum-=(i+1)
    nums.append((max(i,int(sum))))

def cutRod(price, n): 
    val = [0 for x in range(n+1)] 
    val[0] = 0

    for i in range(1, n+1): 
        max_val = -1
        for j in range(i):max_val = max(max_val, price[j] + val[i-j-1]) 
        val[i] = max_val 
  
    return val[n] 

q=int(input())
for _ in range(q):
    n=int(input())
    print(cutRod(nums[1:n+1],n))